flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java

```diff
@@ -729,13 +738,33 @@ private void registerTableSinkInternal(String name, TableSink<?> tableSink) {
 	}
 
 	private void createCatalogFunction(CreateFunctionOperation createFunctionOperation) {
-		Catalog catalog = getCatalogOrThrowException(createFunctionOperation.getFunctionIdentifier().getCatalogName());
 		String exMsg = getDDLOpExecuteErrorMsg(createFunctionOperation.asSummaryString());
 		try {
-			catalog.createFunction(
-				createFunctionOperation.getFunctionIdentifier().toObjectPath(),
-				createFunctionOperation.getCatalogFunction(),
-				createFunctionOperation.isIgnoreIfExists());
+			CatalogFunction function = createFunctionOperation.getCatalogFunction();
+			if (function.isTemporary()) {
+				boolean exist = functionCatalog.hasTemporaryCatalogFunction(
+					createFunctionOperation.getFunctionIdentifier());
+				if (!exist) {
+					FunctionDefinition functionDefinition = FunctionDefinitionUtil.createFunctionDefinition(
+						createFunctionOperation.getFunctionName(),
+						function);
+					registerFunctionInFunctionCatalog(
+						createFunctionOperation.getFunctionIdentifier(),
+						functionDefinition);
+				} else if (!createFunctionOperation.isIgnoreIfExists()) {
+					throw new FunctionAlreadyExistException(FunctionCatalog.TEMPORARY_CATALOG,
+						createFunctionOperation.getFunctionIdentifier().toObjectPath());
+				}
+			} else {
+				Catalog catalog = getCatalogOrThrowException(
+					createFunctionOperation.getFunctionIdentifier().getCatalogName());
+				catalog.createFunction(
```

Comment:
catch FunctionAlreadyExistException and rethrow ValidationException directly?

Reply:
It is the current behavior, right? The function can throw both catalog doesn't exist Validation Exception and also FunctionAlreadyExistException. They are handled in this way, so that validation exception can be thrown without wrapped to a TableException.

Reply:
ok, I see what you mean.